src/cli/keepassxc-shell.1
.TH KEEPASSXC-SHELL 1 "FEB 13, 2019"

.SH NAME
keepassxc-shell \- password shell for the \fBKeePassXC\fP password manager.

.SH SYNOPSIS
.B keepassxc-shell
.I command
.RI [ options ]

.SH DESCRIPTION
\fBkeepassxc-shell\fP is the password shell for the \fBKeePassXC\fP password manager. It provides the ability to query and modify the entries of a KeePass database, directly from the command line.

.SH COMMANDS

.IP "add [options] <entry>"
Adds a new entry to a database. A password can be generated (\fI-g\fP option), or a prompt can be displayed to input the password (\fI-p\fP option).

.IP "clip [options] <entry> [timeout]"
Copies the password or the current TOTP (\fI-t\fP option) of a database entry to the clipboard. If multiple entries with the same name exist in different groups, only the password for the first one is going to be copied. For copying the password of an entry in a specific group, the group path to the entry should be specified as well, instead of just the name. Optionally, a timeout in seconds can be specified to automatically clear the clipboard.

.IP "edit [options] <entry>"
Edits a database entry. A password can be generated (\fI-g\fP option), or a prompt can be displayed to input the password (\fI-p\fP option).

.IP "extract [options]"
Extracts and prints the contents of a database to standard output in XML format.

.IP "locate [options] <term>"
Locates all the entries that match a specific search term in a database.

.IP "lock [options]"
Locks a database.

.IP "ls [options] <database> [group]"
Lists the contents of a group in a database. If no group is specified, it will default to the root group.

.IP "merge [options] <database2>"
Merges the current database with another database.
It is advisable to keep a backup of the two database files before attempting a merge.
In the case that both databases make use of the same credentials, the \fI--same-credentials\fP or \fI-s\fP option can be used.

.IP "rm [options] <entry>"
Removes an entry from a database. If the database has a recycle bin, the entry will be moved there. If the entry is already in the recycle bin, it will be removed permanently.

.IP "show [options] <entry>"
Shows the title, username, password, URL and notes of a database entry. Can also show the current TOTP. Regarding the occurrence of multiple entries with the same name in different groups, everything stated in the \fIclip\fP command section also applies here.

.IP "status [options]"
Show current shell status.

.IP "unlock [options] <database>"
Unlocks a database.

.SH OPTIONS

.SS "General options"

.IP "-h, --help"
Displays help information.

.IP "-v, --version"
Shows the program version.


.SS "Unlock options"

.IP "-k, --key-file <path>"
Specifies a path to a key file for unlocking the database. In a merge operation this option is used to specify the key file path for the first database.

.IP "-q, --quiet <path>"
Silence password prompt and other secondary outputs.


.SS "Merge options"

.IP "-f, --key-file-from <path>"
Path of the key file for the second database.

.IP "-s, --same-credentials"
Use the same credentials for unlocking both database.


.SS "Add and edit options"

.IP "-u, --username <username>"
Specify the username of the entry.

.IP "--url <url>"
Specify the URL of the entry.

.IP "-p, --password-prompt"
Use a password prompt for the entry's password.

.IP "-g, --generate"
Generate a new password for the entry.

.IP "-l, --password-length"
Specify the length of the password to generate.


.SS "Edit options"

.IP "-t, --title <title>"
Specify the title of the entry.


.SS "Clip options"

.IP "-t, --totp"
Copy the current TOTP instead of current password to clipboard. Will report an error
if no TOTP is configured for the entry.


.SS "Show options"

.IP "-a, --attributes <attribute>..."
Names of the attributes to show. This option can be specified more than once,
with each attribute shown one-per-line in the given order. If no attributes are
specified and \fI-t\fP is not specified, a summary of the default attributes is given.

.IP "-t, --totp"
Also show the current TOTP. Will report an error if no TOTP is configured for the
entry.


.SS "List options"

.IP "-R, --recursive"
Recursively list the elements of the group.



.SH REPORTING BUGS
Bugs and feature requests can be reported on GitHub at https://github.com/keepassxreboot/keepassxc/issues.

.SH AUTHOR
This manual page was originally written by Manolis Agkopian <m.agkopian@gmail.com>,
and is maintained by the KeePassXC Team <team@keepassxc.org>.


src/cli/keepassxc-shell.cpp
/*
 *  Copyright (C) 2017 KeePassXC Team <team@keepassxc.org>
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 2 or (at your option)
 *  version 3 of the License.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

#include <cstdlib>
// move this and the fork to cli/Utils
#include <unistd.h>

#include <QCommandLineParser>
#include <QStringList>
#include <QProcess>

#include "cli/ShellApplication.h"
#include "cli/TextStream.h"
#include "cli/List.h"
#include <cli/Command.h>
#include "cli/Utils.h"

#include "config-keepassx.h"
#include "core/Bootstrap.h"
#include "crypto/Crypto.h"

#if defined(WITH_ASAN) && defined(WITH_LSAN)
#include <sanitizer/lsan_interface.h>
#endif

int main(int argc, char** argv)
{
    if (!Crypto::init()) {
        qFatal("Fatal error while testing the cryptographic functions:\n%s", qPrintable(Crypto::errorString()));
        return EXIT_FAILURE;
    }

    ShellApplication app(argc, argv);
    ShellApplication::setApplicationVersion(KEEPASSXC_VERSION);


    Bootstrap::bootstrap();

    TextStream out(stdout);
    TextStream errorTextStream(Utils::STDERR, QIODevice::WriteOnly);
    QStringList arguments;
    for (int i = 0; i < argc; ++i) {
        arguments << QString(argv[i]);
    }

    QString description("KeePassXC database shell.");
    description = description.append(QObject::tr("\n\nAvailable commands:\n"));
    for (Command* command : Command::getCommands()) {
        description = description.append(command->getDescriptionLine());
    }


    QCommandLineParser parser;
    parser.setApplicationDescription(description);
    parser.addPositionalArgument("command", QObject::tr("Name of the command to execute."));

    parser.addHelpOption();
    parser.addVersionOption();
    parser.addOption(Command::QuietOption);
    parser.addOption(Command::KeyFileOption);
    // TODO : use the setOptionsAfterPositionalArgumentsMode (Qt 5.6) function
    // when available. Until then, options passed to sub-commands won't be
    // recognized by this parser.
    parser.parse(arguments);

    if (parser.positionalArguments().empty()) {
        if (parser.isSet("version")) {
            // Switch to parser.showVersion() when available (QT 5.4).
            out << KEEPASSXC_VERSION << endl;
            return EXIT_SUCCESS;
        }
        parser.showHelp();
    }


    QString commandName = parser.positionalArguments().at(0);
    if (commandName == QString("unlock")) {
        if (arguments.size() != 3) {
            errorTextStream << parser.helpText().replace("keepassxc-shell", "keepassxc-shell unlock");
            return EXIT_FAILURE;
        }
        auto db = Utils::unlockDatabase(arguments.at(2),
                                        parser.value(Command::KeyFileOption),
                                        parser.isSet(Command::QuietOption) ? Utils::DEVNULL : Utils::STDOUT,
                                        Utils::STDERR);
        if (!db) {
            return EXIT_FAILURE;
        }
        app.setDatabase(db);
        // TODO print database name instead if populated??
        out << QObject::tr("🔓 Successfully unlocked database %1 🔓").arg(arguments.at(2)) << endl;
        if (!app.isAlreadyRunning()) {

            pid_t processId = fork();
            // parent process. Nothing more to do 🔫.
            if (processId > 0) {
                errorTextStream << QObject::tr("Forked process %1 🍴🍴").arg(processId) << endl;
                app.preserveLock();
                return EXIT_SUCCESS;

            } else if (processId < 0) {
                errorTextStream << QObject::tr("Could not fork 🍴🍴") << endl;
                return EXIT_FAILURE;
            }

            qDebug("yo whats up!!");
            int exitCode = ShellApplication::exec();
            qDebug("daemon killed 🔫 %1", qPrintable(exitCode));
            return exitCode;
        } else {
            return EXIT_SUCCESS;
        }
    }

    if (!app.isAlreadyRunning()) {
        out << QObject::tr("🔑 No database unlocked 🔑") << endl;
        return EXIT_FAILURE;
    }

    app.sendArgumentsToRunningInstance(arguments);

#if defined(WITH_ASAN) && defined(WITH_LSAN)
    // do leak check here to prevent massive tail of end-of-process leak errors from third-party libraries
    __lsan_do_leak_check();
    __lsan_disable();
#endif

    return EXIT_SUCCESS;
}
